Mathlib/Tactic/NormNum/IsSquare.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2025 Vasilii Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasilii Nesterov
+-/
+import Mathlib.Tactic.NormNum.Irrational
+
+/-! # `norm_num` extension for `IsSquare`
+
+This module defines a `norm_num` extension for `IsSquare x` for rational `x`.
+
+## Implementation details
+To decide `IsSquare x`, we reduce it to `0 ≤ x ∧ ¬ Irrational √x` and use the existing `norm_num`
+extensions for `Irrational` and `LE`.
+-/
+
+namespace Tactic.NormNum
+
+section Lemmas
+
+private lemma IsSquare_nat_iff_not_Irrational_sqrt (n : ℕ) :
+    ¬ Irrational √(n : ℝ) ↔ IsSquare n := by
+  simp [irrational_sqrt_natCast_iff]
+
+private lemma IsSquare_int_iff_not_Irrational_sqrt (n : ℤ) (h : 0 ≤ n) :
+    ¬ Irrational √(n : ℝ) ↔ IsSquare n := by
+  rw [irrational_sqrt_intCast_iff]
+  simp [h]
+
+private lemma IsSquare_rat_iff_not_Irrational_sqrt (n : ℚ) (h : 0 ≤ n) :
+    ¬ Irrational √(n : ℝ) ↔ IsSquare n := by
+  rw [irrational_sqrt_ratCast_iff]
+  simp [h]
+
+end Lemmas
+
+open Qq Lean Elab.Tactic Mathlib.Meta.NormNum
+
+/-- `norm_num` extension that decides `IsSquare x` for rational `x`. -/
+@[norm_num IsSquare _]
```

Comment:
I think it would be more idiomatic to have separate extensions for the ℕ, ℤ, and, ℚ cases.

Then there would be a natural way to support other cases in the future (e.g. ZMod n).